### ISSUE TITLE
OSCORE Californium API changes.

### ISSUE THREAD

**Issue by sbernard31** [eclipse/leshan#1231]
I will list here all the Californium changes we talked about with @rikard-sics.

1. probably move classes about `OscoreStore` in [org.eclipse.leshan.core.californium.oscore.cf](https://github.com/eclipse/leshan/tree/oscore/leshan-core-cf/src/main/java/org/eclipse/leshan/core/californium/oscore/cf)(or something like this) from Leshan to Californium.
2. Add a way to specify in EndpointContext the recipient ID targeted. (https://github.com/eclipse/leshan/pull/1212#discussion_r830937966)
3. Being able to create an CoapEnpoint which allow OSCORE communication only.
4. Do not send Request if there is an OSCORE option but no context (see https://github.com/eclipse/leshan/pull/1212#issuecomment-1072266510)

@rikard-sics tell me if I missed something ?

**Comment by rikard-sics** [comment]
Yes I had many of the same points listed down.

The only possible additions are:

**Comment by rikard-sics** [comment]
> It could be better to have an API which doesn't need a CBOR object but just a int ?

In the OSCoreCtx constructor? Yeah I think that would be convenient and make sense.

**Comment by rikard-sics** [comment]
> Yes or/and adding AlgorithmID.fromValue(int value) ?

Yep, that one also makes sense to add.

**Comment by sbernard31** [comment]
(Not really an API changes but maybe a tiny bug ?)
In `org.eclipse.californium.oscore.OSCoreCtx.OSCoreCtx()` line 319
```diff
try {
	this.sender_key = deriveKey(this.common_master_secret, this.common_master_salt, this.key_length, digest,
			info.EncodeToBytes());
} catch (CoseException e) {
-	LOGGER.error(e.getMessage());
-       throw new OSException(e.getMessage());
+	LOGGER.error(e.getMessage(), e); 
+      throw new OSException(e.getMessage(), e);
}
```
First line, it could make sense to display exception stack trace in the log.
Second line, CoseException should be added as cause of OSException

(except all of this is hiding for Security reason ?)

**Comment by rikard-sics** [comment]
> (except all of this is hiding for Security reason ?)

Yes, I did not write that particular code (it was a previous colleague), but I could imagine it was for security reasons.

**Comment by rikard-sics** [comment]
I think we have a good list of changes to be done in Californium then. I have one PR pending related to the plugtest server/client, after that has been merged I can go through and perform these changes.

**Comment by sbernard31** [comment]
I added the point about `Ctx.setURI` raised at : https://github.com/eclipse/leshan/pull/1232#discussion_r841851488

**Comment by sbernard31** [comment]
(@rikard-sics  just to let you know : I will be unavailable from the end of day to Monday 18th april, back on Tuesday)

**Comment by rikard-sics** [comment]
> (@rikard-sics just to let you know : I will be unavailable from the end of day to Monday 18th april, back on Tuesday)

Okay, thanks for letting me know. 

Meanwhile I can spend time on testing the current code (as you mention in #920), and work on implementing the additions to Californium.

**Comment by sbernard31** [comment]
@rikard-sics any news about this :slightly_smiling_face:  ?

**Comment by rikard-sics** [comment]
> @rikard-sics any news about this slightly_smiling_face ?

I have been focusing a bit on the addition of OSCORE support to the sandbox client/server in Californium, and related discussions on interoperability with libcoap that was triggered by this which took some time to investigate.

However, I have begun working on some of the points listed and am planning to create a PR covering those. Currently I am on a business trip and will be back on Saturday so probably I can get this first PR done by the middle of next week. 

When I create the PR you could also review it to check that it suits the intended functionality.

**Comment by sbernard31** [comment]
OK I will be often unavailable during April and May. (e.g. I'm back on Tuesday)

Do you think Leshan opened PR about OSCORE should be merge in `oscore` branch or prefer to wait you test it more ?

**Comment by sbernard31** [comment]
@rikard-sics, I integrated all opened OSCORE PR in `oscore` branch.

For the integration in `master` I don't know yet, maybe too soon.  :thinking: 

Just to let you know about my availability on May : 

I will be unavailable :
- from April 29th to May 8th (included)
- from May 13th to 22th
- and  26/27th

then I will be avaialble again as usual :slightly_smiling_face:.

**Comment by sbernard31** [comment]
@rikard-sics, I try to move forward on Bootstrap config support at server side : #1254 

(to do that I rebase `oscore` branch again on master)

**Comment by sbernard31** [comment]
@rikard-sics, let me know if I can help in any way ?

**Comment by sbernard31** [comment]
Working on #1220, **I found a blocker issue** :x:.
For transport layer abstraction, we need to know the endpoint which receive a request.
We currently retrieve it using `Message.getLocalAddress()` : 
```java
/**
 * Gets the local address of the receiving endpoint.
 * 
 * @return local address of the receiving endpoint. {@code null} for
 *         outgoing messages.
 * @since 3.0
 */
public InetSocketAddress getLocalAddress() {
	return localAddress;
}
```
But with OSCORE this returns `null`, maybe because a request is created under the hood and this field in not set.
(maybe here ? : https://github.com/eclipse/californium/blob/main/cf-oscore/src/main/java/org/eclipse/californium/oscore/OptionJuggle.java#L280-L291)
